Optionally pass the swagger doc into the req object

johndcarmichael opened the pull request:
During development we work with many micro-services. in swagger 2 you can only have a single host which poses a problem during development and the developers was to use the swagger to test their api.

This pr injects the swagger doc into the req object and thus allowing the dev to inject an additional middleware into the mix.. in the readme (and the one I use from the fork), I can modify the swagger host based on the express host within the request object. This dynamically sets the host based on the env the ap is running, local -> dev -> int etc etc.

scottie1984 commented:
Hey - my only issue is that generateHTML is a synchronous function and was not intended for use at run time. Using it like this would therefore block the thread.

The generateHTML function is exported so is using this directly an option https://github.com/scottie1984/swagger-ui-express/issues/65#issuecomment-378280126

Also is modifying the host before its passed into the swagger setup an option?

```
const express = require('express');
const app = express();
const swaggerUi = require('swagger-ui-express');
const swaggerDocument = require('./swagger.json');

swaggerDocument.host = req.get('host');

app.use('/api-docs', swaggerUi.serve, swaggerUi.setup(swaggerDocument));
```

Acrontum-Carmichael commented:
Not really, as the host varies depending on which dev environment is in question. To know the url means the container would need to its url.. which it doesn't.. during dev/stage though performance is not a concern so calling a sync function is a non-issue. So.. you can only dynamically fetch the incoming url during the middleware, or setup some elaborate middleware incoming host name cache thing which seems ott. 

I do see your point though for anyone running this on production it would be problematic. Would an extension to the readme suffice, explaining the pros and cons of each use case?

Acrontum-Carmichael commented:
Alternatively, the tpl files could be passed to the generateHTML function instead bypassing the problem completely.

This would add another avenue for a new feature, "Pass in the location of the tpl files":
- when using webpack to transpile all the code into a single deployable build file, this package breaks this pattern as then the built file looks for the tpl files locally to the build folder.. eg
```
/build/server.js
/node_modules/swagger...
```
By allowing the user to specify the location of the js and html tpl files, this problem goes away.

scottie1984 commented:
Ah yes - makes sense why you want something run time. I am happy with either approach.

Acrontum-Carmichael commented:
I will write into this package then to pass the rendered tpl content to the html generator function.

Looking more into the code base, the package also exposes these options which are not documented:
```
  serveWithOptions: serveWithOptions,
  generateHTML: generateHTML,
  serveFiles: serveFiles
```
Is this for any reason or can this be removed?

johndcarmichael commented:
Well, i updated the pr and readme to pass in the html and js strings to the html generation function.

Also, also updated the syntax to use js standard throughout, the style was a mixture of different styles, removed a non used function and a typo trailing comma

scottie1984 commented on index.js:
```diff
@@ -22,47 +24,47 @@ var generateHTML = function (swaggerDoc, opts, options, customCss, customfavIcon
+  return function (req, res) {
+    res.send(
+      generateHTML(swaggerDoc || req.swaggerDoc, opts, options, customCss, customfavIcon, swaggerUrl, customeSiteTitle, htmlTplString, jsTplString)
+    )
```
can the generateHTML function be called outside the  ```return function (req, res) { ```. It just seems a little overkill to have to regenerate the HTML again with each request. I think having the if statement you had in the previous commit is a nice compromise

scottie1984 commented on README.md:
```diff
@@ -208,6 +208,24 @@ const swaggerDocument = YAML.load('./swagger.yaml');
+const app = express();
+const swaggerUi = require('swagger-ui-express');
+app.use('/api-docs', function(req, res, next){
+    const swaggerDocument = require('./swagger.json');
```
can you place this require outside the ```app.use```? don't want to promote blocking the thread
